### PR TITLE
Breaking circular references in ported input system.

### PR DESCRIPTION
**Pull request by StephenHodgson**
TODO:
- [x] Fix InputManager instantiation in MixedRealityManager.

### CODE REVIEW

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/INavigationHandler.cs [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
 
-using Microsoft.MixedReality.Toolkit.InputSystem.EventData;
+using Microsoft.MixedReality.Toolkit.Internal.EventDatum.Input;
 using UnityEngine.EventSystems;
 
-namespace Microsoft.MixedReality.Toolkit.InputSystem.InputHandlers
+namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem.Handlers
 {
     /// <summary>
     /// Interface to implement to react to navigation gestures.
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/INavigationHandler.cs]: MRTK Interface names should be prefixed with either IMixedReality or IMixedRealityToolkit

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/IPlacementHandler.cs [hunk 2]
```diff
@@ -1,10 +1,10 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
-using Microsoft.MixedReality.Toolkit.InputSystem.EventData;
+using Microsoft.MixedReality.Toolkit.Internal.EventDatum.Input;
 using UnityEngine.EventSystems;
 
-namespace Microsoft.MixedReality.Toolkit.InputSystem.InputHandlers
+namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem.Handlers
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/IPlacementHandler.cs]: MRTK Interface names should be prefixed with either IMixedReality or IMixedRealityToolkit

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/IPointerHandler.cs [hunk 3]
```diff
@@ -1,10 +1,10 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
-using Microsoft.MixedReality.Toolkit.InputSystem.EventData;
+using Microsoft.MixedReality.Toolkit.Internal.EventDatum.Input;
 using UnityEngine.EventSystems;
 
-namespace Microsoft.MixedReality.Toolkit.InputSystem.InputHandlers
+namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem.Handlers
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/IPointerHandler.cs]: MRTK Interface names should be prefixed with either IMixedReality or IMixedRealityToolkit

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/ISourceStateHandler.cs [hunk 4]
```diff
@@ -1,10 +1,10 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
-using Microsoft.MixedReality.Toolkit.InputSystem.EventData;
+using Microsoft.MixedReality.Toolkit.Internal.EventDatum.Input;
 using UnityEngine.EventSystems;
 
-namespace Microsoft.MixedReality.Toolkit.InputSystem.InputHandlers
+namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem.Handlers
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/ISourceStateHandler.cs]: MRTK Interface names should be prefixed with either IMixedReality or IMixedRealityToolkit

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/ISpeechHandler.cs [hunk 5]
```diff
@@ -1,10 +1,10 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
-using Microsoft.MixedReality.Toolkit.InputSystem.EventData;
+using Microsoft.MixedReality.Toolkit.Internal.EventDatum.Input;
 using UnityEngine.EventSystems;
 
-namespace Microsoft.MixedReality.Toolkit.InputSystem.InputHandlers
+namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem.Handlers
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/ISpeechHandler.cs]: MRTK Interface names should be prefixed with either IMixedReality or IMixedRealityToolkit

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/ITeleportHandler.cs [hunk 6]
```diff
@@ -1,10 +1,10 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
-using Microsoft.MixedReality.Toolkit.InputSystem.EventData;
+using Microsoft.MixedReality.Toolkit.Internal.EventDatum.Input;
 using UnityEngine.EventSystems;
 
-namespace Microsoft.MixedReality.Toolkit.InputSystem.InputHandlers
+namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem.Handlers
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/ITeleportHandler.cs]: MRTK Interface names should be prefixed with either IMixedReality or IMixedRealityToolkit

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/ICursor.cs [hunk 7]
```diff
@@ -1,11 +1,10 @@
 ﻿// Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
-using Microsoft.MixedReality.Toolkit.InputSystem.InputHandlers;
-using Microsoft.MixedReality.Toolkit.InputSystem.Pointers;
+using Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem.Handlers;
 using UnityEngine;
 
-namespace Microsoft.MixedReality.Toolkit.InputSystem.Cursors
+namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/ICursor.cs]: MRTK Interface names should be prefixed with either IMixedReality or IMixedRealityToolkit

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/IInputSource.cs [hunk 8]
```diff
@@ -2,10 +2,9 @@
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
 using System.Collections;
-using Microsoft.MixedReality.Toolkit.InputSystem.Pointers;
 using Microsoft.MixedReality.Toolkit.Internal.Definitions;
 
-namespace Microsoft.MixedReality.Toolkit.InputSystem.Sources
+namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/IInputSource.cs]: MRTK Interface names should be prefixed with either IMixedReality or IMixedRealityToolkit

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/IPointer.cs [hunk 9]
```diff
@@ -2,15 +2,11 @@
...
-using Microsoft.MixedReality.Toolkit.InputSystem.Sources;
-using Microsoft.MixedReality.Toolkit.Internal.Interfaces;
+using Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem.Handlers;
+using Microsoft.MixedReality.Toolkit.Internal.Utilities;
 using UnityEngine;
 
-namespace Microsoft.MixedReality.Toolkit.InputSystem.Pointers
+namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem
 {
     public interface IPointer : IEqualityComparer
```
Comment (SimonDarksideJ) [inline Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/IPointer.cs]: MRTK Interface names should be prefixed with either IMixedReality or IMixedRealityToolkit

File: Assets/MixedRealityToolkit/_Core/Managers/MixedRealityManager.cs [hunk 10]
```diff
@@ -70,10 +70,9 @@ private void ResetConfiguration()
         /// <summary>
         /// Local component registry for the Mixed Reality Manager, to allow runtime use of the Manager.
         /// </summary>
-        [HideInInspector]
```
[the remark this answers is not in the document]
Comment (SimonDarksideJ) [reply Assets/MixedRealityToolkit/_Core/Managers/MixedRealityManager.cs]: Still pending :D

[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Managers/MixedRealityManager.cs]: Cause it doesn't show in the inspector anyway.

File: Assets/MixedRealityToolkit-Tests/MixedRealityManagerTests.cs [hunk 11]
```diff
@@ -1,16 +1,21 @@
...
+using System.Linq;
 using UnityEngine;
 using UnityEngine.TestTools;
 
 namespace Microsoft.MixedReality.Toolkit.Tests
 {
     public class MixedRealityManagerTests
     {
-        public void CreateMRManager()
+        public static void CreateMixedRealityManager()
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit-Tests/MixedRealityManagerTests.cs]: Then should the method be private?

File: Assets/MixedRealityToolkit/DefaultProfile.asset [hunk 12]
```diff
@@ -0,0 +1,19 @@
+%YAML 1.1
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/DefaultProfile.asset]: Didn't realize there was one already. I'll be sure to update it with the default settings as well.

File: Assets/MixedRealityToolkit/InputSystem/Focus/FocusProvider.cs [hunk 13]
```diff
@@ -202,24 +202,45 @@ private void Awake()
...
-            if (uiRaycastCamera != null)
-            {
-                UIRaycastCamera = uiRaycastCamera;
-            }
-            else
-            {
-                Debug.LogError("Error initializing UIRaycast Camera!");
-            }
+                Debug.LogWarning("No UIRaycastCamera assigned! Creating a new UIRaycastCamera.");
+                var cameraObject = new GameObject();
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/InputSystem/Focus/FocusProvider.cs]: I could put it in it's own method call, but I think it should stay in the `FocusProvider`. It's rather specific to it.

[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/InputSystem/Focus/FocusProvider.cs]: Besides this is just a fallback if someone didn't set up their own prefab correctly.

File: Assets/MixedRealityToolkit/InputSystem/Sources/InteractionInputSources.cs [hunk 14]
```diff
@@ -364,7 +365,7 @@ public void StopGestureRecognizer()
         /// <returns>True if data is available.</returns>
         public static bool TryGetSourceKind(uint sourceId, out InteractionSourceKind sourceKind)
         {
-            foreach (var inputSource in interactionInputSources)
+            foreach (var inputSource in SourceList)
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/InputSystem/Sources/InteractionInputSources.cs]: This has nothing to do with scene item counts anyway.  The number of input sources will probably never be more than two or four.

File: Assets/MixedRealityToolkit/_Core/EventDatum/GenericBaseEventData.cs [hunk 15]
```diff
@@ -1,18 +1,16 @@
...
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
-using Microsoft.MixedReality.Toolkit.Internal.Interfaces;
+using Microsoft.MixedReality.Toolkit.Internal.Interfaces.Events;
 using UnityEngine.EventSystems;
 
-namespace Microsoft.MixedReality.Toolkit.Internal.Events
+namespace Microsoft.MixedReality.Toolkit.Internal.EventDatum
 {
     public class GenericBaseEventData : BaseEventData
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/EventDatum/GenericBaseEventData.cs]: updated all event documentations.

File: Assets/MixedRealityToolkit/_Core/EventDatum/Input/InputDualAxisPositionEventData.cs [hunk 16]
```diff
@@ -1,12 +1,11 @@
...
-using Microsoft.MixedReality.Toolkit.InputSystem.Sources;
 using Microsoft.MixedReality.Toolkit.Internal.Definitions;
+using Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem;
 using UnityEngine;
-using UnityEngine.EventSystems;
 
-namespace Microsoft.MixedReality.Toolkit.InputSystem.EventData
+namespace Microsoft.MixedReality.Toolkit.Internal.EventDatum.Input
 {
     public class InputDualAxisPositionEventData : InputEventData
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/EventDatum/Input/InputDualAxisPositionEventData.cs]: done

File: Assets/MixedRealityToolkit/_Core/EventDatum/Input/InputPressedEventData.cs [hunk 17]
```diff
@@ -1,13 +1,12 @@
...
 using Microsoft.MixedReality.Toolkit.Internal.Definitions;
+using Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem;
 using UnityEngine;
-using UnityEngine.EventSystems;
 
 
-namespace Microsoft.MixedReality.Toolkit.InputSystem.EventData
+namespace Microsoft.MixedReality.Toolkit.Internal.EventDatum.Input
 {
     public class InputPressedEventData : InputEventData
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/EventDatum/Input/InputPressedEventData.cs]: done

File: Assets/MixedRealityToolkit/_Core/Extensions/InteractionSourceExtensions.cs [hunk 18]
```diff
@@ -118,11 +118,11 @@ public static IAsyncOperation<IRandomAccessStreamWithContentType> TryGetRenderab
                 {
                     IReadOnlyList<SpatialInteractionSourceState> sources = SpatialInteractionManager.GetForCurrentView().GetDetectedSourcesAtTimestamp(PerceptionTimestampHelper.FromHistoricalTargetTime(DateTimeOffset.Now));
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Extensions/InteractionSourceExtensions.cs]: I like it short and sweet.

File: Assets/MixedRealityToolkit/_Core/Interfaces/EventSystem/IMixedRealityEventHandler.cs [hunk 19]
```diff
@@ -1,10 +1,10 @@
...
 // Licensed under the MIT License. See LICENSE in the project root for license information.
 
-using Microsoft.MixedReality.Toolkit.Internal.Events;
+using Microsoft.MixedReality.Toolkit.Internal.EventDatum;
 using UnityEngine.EventSystems;
 
-namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces
+namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces.Events
 {
     public interface IMixedRealityEventHandler : IEventSystemHandler
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/EventSystem/IMixedRealityEventHandler.cs]: done

File: Assets/MixedRealityToolkit/_Core/Interfaces/EventSystem/IMixedRealityEventSource.cs [hunk 20]
```diff
@@ -3,7 +3,7 @@
 
 using System.Collections;
 
-namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces
+namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces.Events
 {
     public interface IMixedRealityEventSource : IEqualityComparer
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/EventSystem/IMixedRealityEventSource.cs]: done

File: Assets/MixedRealityToolkit/_Core/Interfaces/EventSystem/IMixedRealityEventSystem.cs [hunk 21]
```diff
@@ -5,7 +5,7 @@
 using UnityEngine;
 using UnityEngine.EventSystems;
 
-namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces
+namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces.Events
 {
     public interface IMixedRealityEventSystem : IMixedRealityManager
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/EventSystem/IMixedRealityEventSystem.cs]: done

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/INavigationHandler.cs (hunk 1 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/INavigationHandler.cs]: done

File: Assets/MixedRealityToolkit/_Core/Interfaces/EventSystem/IMixedRealityEventSystem.cs (hunk 21 above)
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/EventSystem/IMixedRealityEventSystem.cs]: done

File: Assets/MixedRealityToolkit/_Core/Interfaces/EventSystem/IMixedRealityEventSource.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/EventSystem/IMixedRealityEventSource.cs]: done

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/IPlacementHandler.cs (hunk 2 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/IPlacementHandler.cs]: done

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/IPointerHandler.cs (hunk 3 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/IPointerHandler.cs]: done

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/ISourceStateHandler.cs (hunk 4 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/ISourceStateHandler.cs]: done

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/ISpeechHandler.cs (hunk 5 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/ISpeechHandler.cs]: done

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/ITeleportHandler.cs (hunk 6 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/Handlers/ITeleportHandler.cs]: done

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/ICursor.cs (hunk 7 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/ICursor.cs]: done

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/IFocusProvider.cs [hunk 22]
```diff
@@ -1,16 +1,35 @@
...
+using Microsoft.MixedReality.Toolkit.Internal.Utilities;
 using UnityEngine;
 
-namespace Microsoft.MixedReality.Toolkit.InputSystem.Cursors
+namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem
 {
-    /// <summary>
-    /// Cursor Modifier Interface that provides basic overrides for cursor behavior.
-    /// </summary>
+    public interface IFocusProvider : ISourceStateHandler
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/IFocusProvider.cs]: Going with `IMixedRealtiy...`

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/IGazeProvider.cs [hunk 23]
```diff
@@ -0,0 +1,7 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+namespace Microsoft.MixedReality.Toolkit.Internal.Interfaces.InputSystem
+{
+    public interface IGazeProvider : IInputSource { }
```
[the remark this answers is not in the document]
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/IGazeProvider.cs]: done

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/IInputSource.cs (hunk 8 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/IInputSource.cs]: lol I got it the first time.

File: Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/IPointer.cs (hunk 9 above)
Comment (StephenHodgson) [reply Assets/MixedRealityToolkit/_Core/Interfaces/InputSystem/IPointer.cs]: Copy pasta done.

**Comment by StephenHodgson** [comment]
> All the InputSystem code added in to the Utilities folder, should be in a "Input" subfolder to identify them as Input Utilities.

I moved a few data holding classes, but I don't think they're really specific to input... Not sure how I want to organize that just yet.

**Comment by StephenHodgson** [comment]
Nothing I can think of atm

**Pull request merged**